### PR TITLE
fix expect interface for newer ptyprocess

### LINKED ISSUE #32147: Make sage.interfaces.expect compatible with ptyprocess >= 0.5.2
<div id="comment:0"></div>

(from [#31280 comment:42](https://github.com/sagemath/sage/issues/31280#comment:42))

CC:  @antonio-rojas @isuruf @kiwifb @dimpase @orlitzky

Component: **interfaces**

_Issue created by migration from https://trac.sagemath.org/ticket/32147_

### PR DESCRIPTION
**Pull request by dimpase**
This will close https://github.com/sagemath/sage/issues/32147

ptyprocess 0.5.2+ included a change that broke our expect.py on macOS

Here we guard a bit more around thrown errors, and this appears to suffice

### CODE REVIEW

File: src/sage/interfaces/sagespawn.pyx [hunk 1]
```diff
@@ -199,9 +199,12 @@ class SagePtyProcess(PtyProcess):
...
                     # think we have any non-ASCII quit_strings anyways.
                     self.write(str_to_bytes(self.quit_string))
-                except (OSError, IOError):
+                except (OSError, IOError, PtyProcessError):
                     pass
-            self.fileobj.close()
+            try:
+                self.fileobj.close()
+            except (OSError, IOError, PtyProcessError):
+                pass
```
Comment (tornaria) [inline src/sage/interfaces/sagespawn.pyx]: What is this for?

Comment (dimpase) [reply src/sage/interfaces/sagespawn.pyx]: it's something only making a difference on macOS. Without it, tests fail.

Comment (dimpase) [reply src/sage/interfaces/sagespawn.pyx]: here is what one gets without this change on macOS. On Linux these two commands complete just fine.
```
sage: gap.eval('quit;')
''
sage: gap.quit()
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Cell In [2], line 1
----> 1 gap.quit()

File ~/software/sage/src/sage/interfaces/expect.py:640, in Expect.quit(self, verbose)
    638         else:
    639             print("Exiting %r" % (self._expect,))
--> 640     self._close()
    641 self._reset_expect()

File ~/software/sage/src/sage/interfaces/expect.py:564, in Expect._close(self, force)
    562 try:
    563     if self._expect is not None:
--> 564         self._expect.close(force=force)
    565 except (PtyProcessError, ExceptionPexpect):
    566     self._expect.ptyproc.fd = -1

File /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pexpect/pty_spawn.py:328, in spawn.close(self, force)
    324 self.flush()
    325 with _wrap_ptyprocess_err():
    326     # PtyProcessError may be raised if it is not possible to terminate
    327     # the child.
--> 328     self.ptyproc.close(force=force)
    329 self.isalive()  # Update exit status from ptyproc
    330 self.child_fd = -1

File ~/software/sage/src/sage/interfaces/sagespawn.pyx:204, in sage.interfaces.sagespawn.SagePtyProcess.close()
    202     except (OSError, IOError):
    203         pass
--> 204 self.fileobj.close()
    205 self.fd = -1
    206 self.closed = 1

OSError: [Errno 5] Input/output error
sage: 
```

File: src/sage/interfaces/expect.py [hunk 2]
```diff
@@ -561,7 +562,7 @@ def _close(self, force=True):
         try:
             if self._expect is not None:
                 self._expect.close(force=force)
-        except ExceptionPexpect:
+        except (PtyProcessError, ExceptionPexpect):
```
Comment (tornaria) [inline src/sage/interfaces/expect.py]:
```suggestion
        except (ExceptionPexpect, OSError):
```

Can you try fixing it here instead? At least here it is reasonably documented what is the reason close may fail.

It doesn't seem we need to catch `PtyProcessError` as it is translated to `ExceptionPexpect` by pexpect itself.

Presumably, nothing needs to be changed in `src/sage/interfaces/sagespawn.pyx`.

Comment (dimpase) [reply src/sage/interfaces/expect.py]: OK, this seems to work, too. I'll change the branch.

**Comment by tobiasdiez** [comment]
Also seems to work on conda (as far as this can be checked with the currently broken ci setup)

**Comment by mkoeppe** [comment]
That's a bit quick with the "positive review" here. see https://github.com/sagemath/sage/issues/32147#issuecomment-1736595019

**Review by mkoeppe — APPROVED** [review]
This seems to work well, thanks!

**Pull request merged**